browser/ui/views/pulsing_block/tooltip_webui_bubble_view.cc

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/views/pulsing_block/tooltip_webui_bubble_view.h"
+
+#include <algorithm>
+#include <utility>
+
+// #include "brave/browser/ui/views/sidebar/bubble_border_with_arrow.h"
+#include "brave/browser/ui/webui/private_new_tab_page/brave_private_new_tab_ui.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/base/metadata/metadata_impl_macros.h"
+#include "ui/base/ui_base_types.h"
+#include "ui/base/window_open_disposition.h"
+#include "ui/display/display.h"
+#include "ui/display/screen.h"
+#include "ui/gfx/canvas.h"
+#include "ui/gfx/scoped_canvas.h"
+#include "ui/views/accessibility/accessibility_paint_checks.h"
+#include "ui/views/background.h"
+#include "ui/views/border.h"
+#include "ui/views/bubble/bubble_border.h"
+#include "ui/views/bubble/bubble_border_arrow_utils.h"
+#include "ui/views/bubble/bubble_frame_view.h"
+#include "ui/views/layout/fill_layout.h"
+#include "ui/views/widget/widget_delegate.h"
+#include "ui/views/window/caption_button_layout_constants.h"
+
+namespace {
+// interface paints a border with arrow
+class BorderWithArrow : public views::BubbleBorder {
+ public:
+  explicit BorderWithArrow(Arrow arrow, Shadow shadow, ui::ColorId color_id)
+      : views::BubbleBorder(arrow, shadow, color_id) {
+    set_visible_arrow(true);
+  }
+
+  BorderWithArrow(const BorderWithArrow&) = delete;
+  BorderWithArrow& operator=(const BorderWithArrow&) = delete;
+  ~BorderWithArrow() override = default;
+  void Paint(const views::View& view, gfx::Canvas* canvas) override {
+    // views::BubbleBorder::Paint(view, canvas);
+    // somehow dont trigger PaintNoShadowLegacy at all or override so it doesn't
+    // draw ugly border
+    PaintVisibleArrow(view, canvas);
+  }
+  enum BubbleArrowPart { kFill, kBorder };
+
+ private:
+  // we're copying this from upstream because we only need arrows
+  void PaintVisibleArrow(const views::View& view, gfx::Canvas* canvas) {
```

Comment:
I think it'd be better to remove this copy and make this class firend of BubbleBorder, or make methods you need under protected: section.

Reply:
If we make this class a friend of BubbleBorder which means we need a patch to maintain and if any arrow functionality changes upstream we'll need to manage the changes. The drawn arrow with the above logic works for perfectly with our flat UI aesthetic. Why make methods under protected btw?